### PR TITLE
Fix #2517 : Check for incomplete parentheses to prevent SQL injection.

### LINKED ISSUE #2517: SQL injection in Gorm With using first and find.
### What version of Go are you using (`go version`)?
latest
### Which database and its version are you using?
postgress latest and gorm latest
### Blind Sql injection localhost:8000/user?id=id=1)) or 1=1--
```go
func GetUser(c *gin.Context) {
	var user []models.User
	dbms := db.GetDatabaseConnection() /*  Open connectins */
	defer dbms.Close()
	id := c.Query("id")

	err := dbms.First(&user, id) // Sql Injection in this line /user?id=id=1)) or 1=1--
	if err == nil {
		c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "success", "result": user})

	} else {
		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusBadRequest, "message": "something error", "result": err})
	}
	return

}

```
Need to runnable with [GORM's docker compose config](https://github.com/jinzhu/gorm/blob/master/docker-compose.yml) or please provides your config.

```go
package main

import (
	"fmt"
	"net/http"

	"./db"
	"./models"
	"github.com/gin-gonic/gin"
)

// DBMigrate for Auto Migrate
func DBMigrate() { /* Auto Migrations */
	fmt.Println("[::] Migration Databases .....")
	dbms := db.GetDatabaseConnection() /* Get connction to database */
	defer dbms.Close()
	dbms.AutoMigrate(&models.User{})
	// db.AutoMigrate(&models.Profile{}) /* Migration Models */
	fmt.Println("[::] Migration Databases Done")
}

func main() {
	DBMigrate()
	router := gin.Default()
	router.POST("/user", CreateUser)
	router.GET("/user", GetUser)
	router.Run(":8080")
}

// CreateUser function
func CreateUser(c *gin.Context) {
	var (
		user models.User
	)

	c.BindJSON(&user)
	dbms := db.GetDatabaseConnection() /*  Open connectins */
	defer dbms.Close()
	dbms.Create(&user)
	c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "success", "result": user})
	return
}

// GetUser function
func GetUser(c *gin.Context) {
	var user []models.User
	dbms := db.GetDatabaseConnection() /*  Open connectins */
	defer dbms.Close()
	id := c.Query("id")

	err := dbms.First(&user, id) // Sql Injection in this line /user?id=id=1)) or 1=1--
	if err == nil {
		c.JSON(http.StatusOK, gin.H{"status": http.StatusOK, "message": "success", "result": user})

	} else {
		c.JSON(http.StatusBadRequest, gin.H{"status": http.StatusBadRequest, "message": "something error", "result": err})
	}
	return

}

```

### PR DESCRIPTION
**Pull request by herpiko**
### What did this pull request do?

Fix for #2517

### CODE REVIEW

File: scope.go [hunk 1]
```diff
@@ -277,6 +277,25 @@ func (scope *Scope) AddToVars(value interface{}) string {
 	return scope.Dialect().BindVar(len(scope.SQLVars))
 }
 
+// IsCompleteParentheses check if the string has complete parentheses to prevent SQL injection
+func (scope *Scope) IsCompleteParentheses(value string) bool {
+	i := 0
```
[the remark this answers is not in the document]
Comment (herpiko) [reply scope.go]: Thanks!

**Comment by herpiko** [comment]
@emirb I've applied your comment. Can you check it again please?

**Comment by emirb** [comment]
Thanks @herpiko.

**Comment by wahyuhadi** [comment]
hi All
this issue still exists

**Comment by sneko** [comment]
@wahyuhadi you say the issue is still existing. Could you please provide more information or an example of the code that is vulnerable please?

Thank you,

**Comment by wahyuhadi** [comment]
@sneko
You can see in here
https://github.com/wahyuhadi/gorm-sqlInjection

**Comment by maeglindeveloper** [comment]
@wahyuhadi ,

I see that you are passing, as a primary key, an interface from gin.Context.
When querying with **First** using **inline conditions,** it is said that it only works with integer primary key 👍 

Since there is no prepare statement in this usecase, it seems logical that there is an issue in this example, wdyt ?
Maybe I'm wrong.

```
// Get by primary key (only works for integer primary key)
db.First(&user, 23)
//// SELECT * FROM users WHERE id = 23;
// Get by primary key if it were a non-integer type
db.First(&user, "id = ?", "string_primary_key")
//// SELECT * FROM users WHERE id = 'string_primary_key';

// Plain SQL
db.Find(&user, "name = ?", "jinzhu")
//// SELECT * FROM users WHERE name = "jinzhu";

db.Find(&users, "name <> ? AND age > ?", "jinzhu", 20)
//// SELECT * FROM users WHERE name <> "jinzhu" AND age > 20;

// Struct
db.Find(&users, User{Age: 20})
//// SELECT * FROM users WHERE age = 20;

// Map
db.Find(&users, map[string]interface{}{"age": 20})
//// SELECT * FROM users WHERE age = 20;
```

But still, it seems necessary to use the specific gorm pattern to avoid SQL injection.
I do have a doubt as well regarding the strenght of the library either regarding that topic.

@jinzhu maybe you can lightened us a bit more ? :)

**Comment by wahyuhadi** [comment]
@maeglindeveloper

The injection id=id=1)) or 1=1-- , i call this double param injection

If you change the injection id=email=some@email.com)) or 1=1--

The query will change with where email=some@email.con

**Comment by maeglindeveloper** [comment]
@wahyuhadi yes, I understood first time ;)

**Pull request merged**